File: codeforces/1600~1699/1672/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<int> a(n);
  for (int i = 0; i < n; ++i) cin >> a[i];
  vector<int> b(n);
  for (int i = 0; i < n; ++i) cin >> b[i];

  map<int,int> pool;
  int i = n-1, j = n-1;
  while (i >= 0 && j >= 0) {
    if (j+1 < n && b[j+1] == b[j]) {
      pool[b[j]]++;
      --j;
      continue;
    }

    while (i > 0 && a[i] != b[j])  {
      if (pool[a[i]] == 0) {
        cout << "NO" << endl;
        return;
      }

      pool[a[i]]--;
      --i;
    }
    --i;
    --j;
  }

  cout << "YES" << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1600~1699/1672/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeMap;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let n = read_int::<i32>();
    let a = read_vec::<i32>();
    let b = read_vec::<i32>();

    let mut pool = BTreeMap::new();
    let mut i = (n-1) as i32;
    let mut j = (n-1) as i32;
    while j >= 0 && i >= 0 {
        while j >= 0 && j+1 < n && b[(j+1) as usize] == b[j as usize] {
            pool.entry(b[j as usize]).and_modify(|count| *count += 1).or_insert(1);
            j -= 1;
        }

        // writeln!(writer, "{} {}", i, j).ok();
        while j >= 0 && i >= 0 && a[i as usize] != b[j as usize] {
            let count = pool.get(&a[i as usize]);
            if count.is_none() || *count.unwrap() == 0 {
                writeln!(writer, "NO").ok();
                return;
            }

            pool.entry(a[i as usize]).and_modify(|count| *count -= 1);
            i -= 1;
        }

        i -= 1;
        j -= 1;
    }

    writeln!(writer, "YES").ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
